state.go

```diff
@@ -214,50 +220,59 @@ type SecretState struct {
 	// Lease duration, in seconds, if any when the secret was read
 	LeaseDuration int `json:"lease_duration,omitempty"`
 
-	// TTL, in seconds, if any when the secret was read
-	TTL int `json:"ttl,omitempty"`
-
 	// Secret will be renewed after this portion of its life has passed
 	DurationRatio float64 `json:"duration_ratio,omitempty"`
 
 	// If the secret has no expiration data, don't try to update it
 	DisableAutoUpdate bool `json:"disable_auto_uptdate,omitempty"`
 
 	// Actual secret
-	Data map[string]interface{} `json:"data,omitempty"`
+	Data SecretData `json:"data,omitempty"`
 
 	// Files using this secret
 	FilesUsing JSONSortedStringList `json:"files_using,omitempty"`
 }
 
-func (s *SecretState) TimeToUpdate() time.Duration {
+func (s *SecretState) Ratio() float64 {
 	ratio := s.DurationRatio
 	if ratio == 0 {
 		ratio = DefaultSecretDurationRatio
 	}
+	return ratio
+}
 
-	// Next update for the secret will be based on these rules:
-	// - If we have both a TTL and a lease duration, we use the minimal of them
-	// - If we have only a TTL or a lease duration, we take it
-	// - If we don't have anything, we won't try to update this secret
-	var duration int
-	switch {
-	case s.TTL > 0 && s.LeaseDuration > 0:
-		if s.TTL < s.LeaseDuration {
-			duration = s.TTL
-		} else {
-			duration = s.LeaseDuration
+func (s *SecretState) TTL() (int, bool) {
+	if s.Data == nil {
+		return 0, false
+	}
+	switch ttlNumber := s.Data["ttl"].(type) {
+	case json.Number:
+		ttl, err := ttlNumber.Int64()
+		if err != nil {
+			return 0, false
 		}
-	case s.TTL > 0:
-		duration = s.TTL
-	case s.LeaseDuration > 0:
-		duration = s.LeaseDuration
-	default:
-		// Unknown TTU
-		return 0
+		return int(ttl), true
+	case int:
+		return ttlNumber, true
+	case int64:
+		return int(ttlNumber), true
 	}
+	return 0, false
+}
 
-	return (time.Duration(float64(duration)*ratio) * time.Second) - time.Now().Sub(s.Timestamp)
+func (s *SecretState) TimeToUpdate() (minTTU time.Time, known bool) {
+	for _, source := range secretTTUSources {
+		ttu, err := source(s)
+		if err != nil {
+			log.Printf("Error trying to obtain TTU for secret '%s': %s", s.Name, err)
+			continue
+		}
+		if ttu != nil && (!known || ttu.Before(minTTU)) {
+			minTTU = *ttu
+			known = true
+		}
+	}
+	return
```

Comment:
According to the function, it should return a tuple `(minTTU time.time, known bool)`

Reply:
Golang allows to have [named return values](https://tour.golang.org/basics/7), this has two implications:

1. Variables can be used as if they would have been defined with var in the first line
2. A return without values returns eactly these variables

So in this case it'd be equivalent to:
```
func (s *SecretState) TimeToUpdate() (time.Time, bool) {
    var minTTU time.Time, known bool
    ...
    return minTTU, known
}
```

I think it is useful in some cases, specially to be more explicit with the meaning of the returned values.

Reply:
I see... I didn't know about this feature.
Thanks! :)